src/charm.py

```diff
@@ -0,0 +1,465 @@
+
+    This class will provide the charm with the necessary data to connect to the DB as
+    well as the current relation status.
+    """
+
+    DATABASE_KEY = "topic"
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str,
+    ):
+        super().__init__(charm, relation_name)
+        self.consumer_prefix = f"{charm.model.name}-{charm.app.name}-benchmark-consumer"
+        # We can have only one reference to a Requires object.
+        # Hence, we will store it here:
+        self._internal_client = KafkaRequires(
+            self.charm,
+            self.relation_name,
+            TOPIC_NAME,
+            extra_user_roles="admin",
+            consumer_group_prefix=self.consumer_prefix,
+        )
+
+    @property
+    @override
+    def state(self) -> RelationState:
+        """Returns the state of the database."""
+        if not (
+            self.relation and self.client and self.relation.id in self.client.fetch_relation_data()
+        ):
+            logger.error("Relation data not found")
+            # We may have an error if the relation is gone but self.relation.id still exists
+            # or if the relation is not found in the fetch_relation_data yet
+            return KafkaDatabaseState(
+                self.charm.app,
+                None,
+            )
+        return KafkaDatabaseState(
+            self.charm.app,
+            self.relation,
+            data=self.client.fetch_relation_data()[self.relation.id],
+        )
+
+    @property
+    @override
+    def client(self) -> Any:
+        """Returns the data_interfaces client corresponding to the database."""
+        return self._internal_client
+
+    def bootstrap_servers(self) -> str | None:
+        """Return the bootstrap servers."""
+        return self.state.get().hosts
+
+    def tls(self) -> tuple[str, str] | None:
+        """Return the TLS certificates."""
+        if not self.state.tls_ca:
+            return self.state.tls, None
+        return self.state.tls, self.state.tls_ca
+
+
+class KafkaPeersRelationHandler(PeerRelationHandler):
+    """Listens to all the peer-related events and react to them."""
+
+    @override
+    def peers(self) -> list[str]:
+        """Return the peers."""
+        return [
+            f"{self.relation.data[u]['ingress-address']}:{8080 + 2 * port}"
+            for u in list(self.units()) + [self.this_unit()]
+            for port in range(0, self.charm.config.get("parallel_processes"))
+        ]
+
+
+class KafkaConfigManager(ConfigManager):
+    """The config manager class."""
+
+    def __init__(
+        self,
+        workload: WorkloadBase,
+        database: KafkaDatabaseRelationHandler,
+        peer: KafkaPeersRelationHandler,
+        config: dict[str, Any],
+        labels: Optional[str] = "",
+    ):
+        self.workload = workload
+        self.workload.worker_params_template = KAFKA_WORKER_PARAMS_TEMPLATE
```

Comment:
**big suggestion** I would NEVER EVER modify attribute for an external class within a function. It makes the code very brittle and hard to understand. Imagine from an outside developer

```
my_class = MyClass(a=1)
my_class.a # equals to 1
my_agg = my_business_logic(my_class)
print(my_class.a) # I suppose everyone reading the code would assume my_class.a would still be equal to 1
```

Who modified the class attribute?